### DISCUSSION TITLE
Get server request when available

### DISCUSSION THREAD

**Discussion by danlupascu** [vercel/next.js#17246] [Help] [answered]
Hello.

First of all, thank you to everyone working on Next.js, you're doing a great job!

I was wondering if there's a way to get the server request (when it's available - when the requested page is rendered on the server).

I know that I can get it using `getInitialProps`, but this will disable automatic static optimization.

Thank you!

**Comment by jamesmosier** [comment]
Hi @danlupascu 👋  It sounds like you could use `getServerSideProps` to get request information server side. https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

```js
export async function getServerSideProps(context) {
  const request = context.req;
  return {
    props: {}, // will be passed to the page component as props
  }
}
```

**Reply by danlupascu** [reply]
@jamesmosier thank you for the suggestion, but I can't use it.

Here's what the [docs](https://nextjs.org/docs/basic-features/data-fetching#only-runs-on-server-side) are saying:

> When you request this page on client-side page transitions through next/link (documentation) or next/router (documentation), Next.js sends an API request to the server, which runs getServerSideProps. It’ll return JSON that contains the result of running getServerSideProps, and the JSON will be used to render the page. All this work will be handled automatically by Next.js, so you don’t need to do anything extra as long as you have getServerSideProps defined.


If I use it, then on every page visit, a request will be sent to the API server to get the server side props, which is something I don't need.

**Comment by lfades** [comment, marked as answer]
It's not possible to get access to request info (aka runtime data) using static pages that are always generated at build time.

If you need to access something from the request then using `getServerSideProps` is the way to go as @jamesmosier said. In other words, you can't have `automatic static optimization` enabled for pages with pre-rendering requirements on an incoming request.

**Reply by danlupascu** [reply]
@lfades I totally get and understand that.

I think that you didn't understand my question.

I want to have access to the request in the `_app` when the page is rendered on the server, without opting out of automatic static optimization.

So, in `_app` I would like to get the `request` IF the current page is rendered on the server. If it's not then it's fine too, because I can get the data I want from the client.

If I use getInitialProps then ASO will be turned off for all the app.

**Reply by danlupascu** [reply]
The bottom line is this - the `_app` component will wrap all the pages.

Some of those pages will be rendered on the server and some of them on the client.

I have a global layout in `_app` and I need to know if the user is using a mobile device or not.

In the `_app`, for the server side rendered pages, I would like to get the `request` somehow and in this way to get the user device.

I thought that maybe the `server` sends the `request` to the `_app` component and I can access it somehow.

**Reply by danlupascu** [reply]
I'm not very familiar with the source code, but I think that this line renders the `_app` on the server side: 

https://github.com/vercel/next.js/blob/f3eadac467ba81a22285f918bdad5ada4db2cfc3/packages/next/next-server/server/render.tsx#L829

Why not sending the `ctx` as well here? If the page is auto exported then we won't have access to it. If it's rendered on the server side then we'll have access to it.

**Reply by lfades** [reply]
@danlupascu That's currently not possible. You could achieve something similar on a per-page basis using an utility function on the pages that have access to the request.

**Reply by danlupascu** [reply]
Yes, I was thinking about it too.

I'll create a global function that will return some data based on the context and return the data as page props in all the pages that will use SSR. 

I hope maybe someday there'll be a better approach.

Thank you for your help!

**Reply by parkjoon** [reply]
> @danlupascu That's currently not possible. You could achieve something similar on a per-page basis using an utility function on the pages that have access to the request.

Hey @lfades - Just curious, is this "currently not possible" because of an underlying blocker or just hasn't been implemented yet?

> I'm not very familiar with the source code, but I think that this line renders the `_app` on the server side:
> 
> https://github.com/vercel/next.js/blob/f3eadac467ba81a22285f918bdad5ada4db2cfc3/packages/next/next-server/server/render.tsx#L829
> 
> Why not sending the `ctx` as well here? If the page is auto exported then we won't have access to it. If it's rendered on the server side then we'll have access to it.

^ That seems straightforward to pass the `ctx` value down. Is there something that could break as a result?

**Reply by lfades** [reply]
@parkjoon It could be possible and something we can consider doing in the future. What I really mean is that the team is working on other things and can't look at this for now (look at our current RFCs like i18n and image optimization).

There's always the option for the community to start a RFC and then create an implementation we can review though 👍

**Comment by parkjoon** [comment]
For those still looking for an alternate solution from the ones mentioned in this thread, I found that if I use a custom app and wrap the `getInitialProps` of my individual pages in an HOC (eg. `withXYZContext`), then I can pass whatever I need from `req` in the aforementioned HOC in the return value, which will consequently be passed to `pageProps` in your custom app.

Conceptually, I am just creating an HOC around the usual `getInitialProps`, which has access to `ctx.req` to pass it through as part of the return value of `getInitialProps`, thus making it accessible from the custom app. From the custom app, you can pass `pageProps.req` as a context value, thus making it accessible to all your child pages and components. This lets you avoid:
1) Disabling automatic static optimization by defining `getInitialProps` at the custom app level (which disables it for all pages of that app)
2) Setting the `req` as a context value individually in every page that needs it

Edit:
Regarding point 1), please see clarification: https://github.com/vercel/next.js/discussions/17246#discussioncomment-91427

**Reply by lfades** [reply]
> Disabling automatic static optimization

Every page with `getInitialProps` will have automatic optimization disabled, if it's `_app.js` then it will happen for all pages.

**Reply by parkjoon** [reply]
Yes apologies, I meant to say "Disabling automatic static optimization at the custom app level (which disables it for all pages of that app)". Thank you for the clarification @lfades

**Reply by lfades** [reply]
@parkjoon it still starts with "this lets you avoid:", but you can't avoid disabling automatic optimization in this case.

**Reply by parkjoon** [reply]
Can you elaborate on this? All I am doing is adding `req` as a `pageProp` value in an HOC that's used with `getInitialProps` at the page level.

getInitialProps HOC:

```
const getInitialPropsHOC = (getInitialProps) => async (ctx) => {
  return {
    ...(await getInitialProps(ctx)),
    req: ctx.req
  };
}
```

At the page level, we use:

```
const MyPage = () => { .... };
MyPage.getInitialProps = getInitialPropsHOC((ctx) => {
  const initialData = .... // do my API calls
  return { initialData };
});
```

In the custom app, I don't specify any `getInitialProp` (or similar) at all. I just take in `pageProps` (as a prop) and create a React context provider with the `req` value in the return value of the custom app.

Custom app:

```
const MyApp = (props) => {
  // pageProps.req is available because we set it as a pageProp in "getInitialPropsHOC"
  return (
    <MyContext.Provider value={props.pageProps.req}>
      <RestOfTheAppThatCanNowAccessReqFromContext />
    <MyContext.Provider />
  );
};
// MyApp.getInitialProps = ....; <- not needed here
```

**Reply by lfades** [reply]
@parkjoon In your case, pages that don't use `getInitialProps` will be static, because the custom app doesn't have `getInitialProps`. But any page that has `getInitialProps` will be a serverless function, because static pages can't know what `req` and `res` are.

**Reply by parkjoon** [reply]
Ah yes, yes, totally agree with you. Forgot to mention that assumption in my original post. Thanks for the awesome followup @lfades!
